### PR TITLE
Allow building only specified directories

### PR DESCRIPTION
**Pull request by jakemac53**
Work towards https://github.com/dart-lang/build/issues/539.

Sending this out before adding tests to get general agreement on the strategy first.

Specifically, this has interesting interactions with the `-o` argument, and also somewhat with the `serve` positional args. I chose to resolve these using these rules:

- If `-o` is provided with a specific output directory, add that to the list of directories to build. So for example `pbr build web -o test:build` would build both `web` and `test`, but the output directory would only contain the hoisted files from `test`.
  - The possible weird edge case here is if you do this without a specified directory for -o `pbr build web -o build` that will still only build the `web` directory, but it won't hoist it in the output directory. I think this roughly correlates to the desired behavior in that case, but it might be a little surprising for some people.
- For `serve` it is more straightforward, I simply treat the serve targets as build directories. The only real difference is in serve mode you are allowed to add a port modifier. The `-o` argument is treated the same as above, so `pbr serve web:8080 -o test:build` would build both `web` and `test`, serve `web` on 8080, and create an output directory with the hoisted `test` directory.

### CODE REVIEW

**Comment by jakemac53** [comment]
Note that I do need to figure out what to do with the test arg parser still as well, currently there are some weird interactions there.

**Comment by natebosch** [comment]
@kevmoo - can you take a look at the comment above regarding how args are handled?

File: build_runner/lib/src/entrypoint/options.dart [hunk 1]
```diff
@@ -521,6 +557,22 @@ class _TestCommand extends _BuildRunnerCommand {
       'Performs a single build on the specified targets and then runs tests '
       'using the compiled assets.';
 
+  @override
+  _SharedOptions _readOptions({Iterable<String> defaultBuildDirs}) {
+    // This command doesn't allow specifying directories to build, instead it
+    // always builds the `test` directory.
+    if (argResults.rest.contains('--')) {
```
Comment (natebosch) [inline build_runner/lib/src/entrypoint/options.dart]: Why does including a directory to build mean that `'--'` shows up in `argResults.rest`?

File: build_runner/lib/src/entrypoint/options.dart [hunk 2]
```diff
@@ -649,6 +702,15 @@ class _CleanCommand extends Command<int> {
   }
 }
 
+Set<String> _buildDirsFromOutputMap(Map<String, String> outputMap) {
+  var dirs = new Set<String>();
+  outputMap.forEach((k, v) {
```
Comment (natebosch) [inline build_runner/lib/src/entrypoint/options.dart]: we aren't using `k` here so we could use a for loop around `outputMap.values`. Even better we could express this as what it is rather than how we get it.

```dart
Set<String> _buildDirsFromOutputMap(Map<String, String> outputMap) =>
    outputMap.values.where((v) => v != null).toSet();
```

File: build_runner/lib/src/generate/build_impl.dart [hunk 3]
```diff
@@ -335,6 +341,12 @@ class _SingleBuild {
       if (phase.isOptional) continue;
       await _performanceTracker.trackBuildPhase(phase, () async {
         if (phase is InBuildPhase) {
+          // Bail early if building a non-root package and we are building
```
Comment (natebosch) [inline build_runner/lib/src/generate/build_impl.dart]: do we need to do something to convert a non-optional phase action into something that can be run lazily in this case?

Comment (jakemac53) [reply build_runner/lib/src/generate/build_impl.dart]: this is done btw, going to add a test for and check the logic for the optional outputs tracker as well.

**Comment by jakemac53** [comment]
Note the behavior demonstrated by the test https://github.com/dart-lang/build/pull/1416/files#diff-9f6a3d2b7986c5450eac408bb5fbf963R210.

Since the OptionalOutputsTracker always allows source files you still get multiple directories - but they won't be built, which is a little bit weird.

Also note that you don't get the generated files under the `packages` dir that weren't inputs to an output in the output dir. We could decide to still always generate all files under `lib` directories but I don't think that is actually what you want (for ddc, or anything consuming summaries/kernel files, it certainly isn't).

File: build_runner/lib/src/generate/build_impl.dart (hunk 3 above)
Comment (jakemac53) [reply build_runner/lib/src/generate/build_impl.dart]: this is done as well, PTAL

**Comment by jakemac53** [comment]
Overview of the differences in builds `_test` package when building in release/debug and web/test/both. Note that each of these has an overhead of `31 items, totalling 58.4 MB` for the `packages/$sdk` directory.

- `- o build`:
  - 14.7s with 399 outputs (2099 actions)
  - 2,063 items (and 2 hidden), totalling 78.5 MB
- `- o build --release`:
  - 47.7s with 230 outputs (1863 actions)
  - 134 items (and 2 hidden), totalling 63.5 MB
- `-o web:build`:
  - 8.0s with 31 outputs (1523 actions)
  - 1,852 items (and 2 hidden), totalling 76.1 MB
- `-o web:build --release`:
  - 11.3s with 19 outputs (1505 actions)
  - 102 items (and 2 hidden), totalling 60.3 MB
- `-o test:build`:
  - 14.2s with 378 outputs (2069 actions)
  - 2,032 items (and 2 hidden), totalling 78.4 MB
- `-o test:build --release`:
  - 45.8s with 214 outputs (1841 actions)
  - 120 items (and 2 hidden), totalling 63.2 MB

File: build_runner/test/entrypoint/run_test.dart [hunk 4]
```diff
@@ -16,41 +22,113 @@ main() {
...
+          process.kill();
+          await process.exitCode;
+          return ExitCode.success.code;
+        } else if (nextLine.contains('failed after')) {
+          process.kill();
+          await process.exitCode;
+          return 1;
+        }
+      }
+      // Shouldn't ever reach here.
```
Comment (natebosch) [inline build_runner/test/entrypoint/run_test.dart]: maybe throw?

File: build_runner/lib/src/asset_graph/optional_output_tracker.dart [hunk 5]
```diff
@@ -23,36 +23,44 @@ import 'node.dart';
 class OptionalOutputTracker {
   final _checkedOutputs = <AssetId, bool>{};
   final AssetGraph _assetGraph;
+  final List<String> _buildDirs;
   final List<BuildPhase> _buildPhases;
+  final String _rootPackage;
 
-  OptionalOutputTracker(this._assetGraph, this._buildPhases);
+  OptionalOutputTracker(
```
Comment (natebosch) [inline build_runner/lib/src/asset_graph/optional_output_tracker.dart]: Maybe expand or rephrase the doc comment on line 18?

Comment (jakemac53) [reply build_runner/lib/src/asset_graph/optional_output_tracker.dart]: done

**Comment by jakemac53** [comment]
@natebosch See the change I had to make to this test https://github.com/dart-lang/build/pull/1416/commits/9a473a0f657606e1a7ae0541cc84058869f6b937

This actually is making me rethink the logic a little bit, perhaps we should always build anything that is "required" that is in any `lib` directory? For instance somebody using Sass without angular wouldn't end up getting the compiled css from other packages (or the current package) except for the sass files under the top level dir you are building. That doesn't seem right.

**Comment by jakemac53** [comment]
I went ahead and changed the logic here to always run required phases for `lib/` inputs, @natebosch PTAL

File: build_runner/lib/src/util/util.dart [hunk 6]
```diff
@@ -0,0 +1,12 @@
+import 'package:build/build.dart';
```
Comment (natebosch) [inline build_runner/lib/src/util/util.dart]: I'm not a fan of `util.dart` (I'm fine with a directory named `util` though). Maybe `build_dirs.dart`?

**Comment by kevmoo** [comment]
You did it!
